app/controllers/courses_controller.rb

```diff
@@ -5,25 +5,59 @@
 
 class CoursesController < ApplicationController
   # you need to be able to pick a course to be authorized for it
-  skip_before_action :authorize_user_for_course, only: [:index, :new, :create]
+  skip_before_action :authorize_user_for_course, only: [:index, :show, :new, :create, :join]
   # if there's no course, there are no persistent announcements for that course
-  skip_before_action :update_persistent_announcements, only: [:index, :new, :create]
+  skip_before_action :update_persistent_announcements, only: [:index, :new, :create, :join]
 
   def index
     courses_for_user = User.courses_for_user current_user
 
-    if courses_for_user.any?
+    if not courses_for_user.nil?
       @listing = categorize_courses_for_listing courses_for_user
     else
-      redirect_to(home_no_user_path) && return
+      @listing = []
     end
 
-    render layout: "home"
+    # render layout: "home"
   end
 
-  action_auth_level :show, :student
+  action_no_auth :show
   def show
-    redirect_to course_assessments_url(@course)
+
+    course_name = params[:name]
+    @course = Course.find_by(name: course_name) if course_name
+
+    unless @course
+      redirect_to(controller: :home, action: :error) && return
+    end
+
+    if not current_user.nil?
+
+      cud = @course.course_user_data.find_by(user_id: current_user.id)
+
+      if not cud.nil?
+        redirect_to course_assessments_path(@course)
```

Comment:
I think you need to wrap the URL in parentheses. Otherwise, the url gets combined with the return before the redirect is called.